electrumsv/restapi.py

```diff
@@ -0,0 +1,227 @@
+
+
+def get_app_state():
+    # to monkeypatch app_state in tests
+    return app_state
+
+
+def get_network_type():
+    app_state = get_app_state()
+    if app_state.config.get('testnet'):
+        return 'test'
+    if app_state.config.get('scalingtestnet'):
+        return 'stn'
+    else:
+        return 'main'
+
+
+class Errors:
+    # http 400 bad requests
+    GENERIC_BAD_REQUEST_CODE = 40000
+    URL_INVALID_NETWORK_CODE = 40001
+    URL_NETWORK_MISMATCH_CODE = 40002
+
+    # http 401 unauthorized
+    AUTH_CREDENTIALS_MISSING_CODE = 40102
+    AUTH_UNSUPPORTED_TYPE_CODE = 40103
+
+    # http 402 - 402xx series
+    # http 403 - 403xx series
+    AUTH_CREDENTIALS_INVALID_CODE = 40301
+
+    # http 500 internal server error
+    GENERIC_INTERNAL_SERVER_ERROR = 50000
+
+    AUTH_CREDENTIALS_INVALID_MESSAGE = "Authentication failed (bad credentials)."
+    AUTH_CREDENTIALS_MISSING_MESSAGE = "Authentication failed (missing credentials)."
+    AUTH_UNSUPPORTED_TYPE_MESSAGE = "Authentication failed (only basic auth is supported)."
+    URL_INVALID_NETWORK_MESSAGE = "Only {} networks are supported. You entered: '{}' network."
+    URL_NETWORK_MISMATCH_MESSAGE = "Wallet is on '{}' network. You requested: '{}' network."
+
+
+class Fault(Exception):
+    """Restapi error class"""
+
+    def __init__(self, code=Errors.GENERIC_BAD_REQUEST_CODE, message='Server error'):
+        self.code = code
+        self.message = message
+
+    def __repr__(self):
+        return "Fault(%s, '%s')" % (self.code, self.message)
+
+
+def bad_request(code: int, message: str) -> web.Response:
+    response_obj = {'code': code,
+                    'message': message}
+    return web.json_response(data=response_obj, status=400)
+
+
+def unauthorized(code: int, message: str) -> web.Response:
+    response_obj = {'code': code,
+                    'message': message}
+    return web.json_response(data=response_obj, status=401)
+
+
+def forbidden(code: int, message: str) -> web.Response:
+    response_obj = {'code': code,
+                    'message': message}
+    return web.json_response(data=response_obj, status=403)
+
+
+def not_found(code: int, message: str) -> web.Response:
+    response_obj = {'code': code,
+                    'message': message}
+    return web.json_response(data=response_obj, status=404)
+
+
+def internal_server_error(code: int, message: str) -> web.Response:
+    response_obj = {'code': code,
+                    'message': message}
+    return web.json_response(data=response_obj, status=500)
+
+
+def good_response(response: Dict) -> web.Response:
+    return web.Response(text=json.dumps(response, indent=2), content_type="application/json")
+
+
+async def decode_request_body(request) -> Union[Dict[Any, Any], Fault]:
+    body = await request.read()
+    if body == b"" or body == b"{}":
+        return {}
+    return json.loads(body.decode('utf-8'))
+
+
+def fault_to_http_response(fault: Union[Fault]):
```

Comment:
A `Union` is saying the typing can be one of a selection of types. If you are just using a base or abstract class as a stand-in for all the subclasses, then you should likely remove the `Union`.

Reply:
Ah yes. It used to be Union[SessionError, Fault] but forgot to remove Union when it was down to one type.